libraries/chain/include/eosio/chain/webassembly/interface.hpp

```diff
@@ -0,0 +1,581 @@
+         int32_t recover_key(legacy_ptr<const fc::sha256>, legacy_array_ptr<char>, legacy_array_ptr<char>) const;
+         void assert_sha256(legacy_array_ptr<char>, legacy_ptr<const fc::sha256>) const;
+         void assert_sha1(legacy_array_ptr<char>, legacy_ptr<const fc::sha1>) const;
+         void assert_sha512(legacy_array_ptr<char>, legacy_ptr<const fc::sha512>) const;
+         void assert_ripemd160(legacy_array_ptr<char>, legacy_ptr<const fc::ripemd160>) const;
+         void sha256(legacy_array_ptr<char>, legacy_ptr<fc::sha256>) const;
+         void sha1(legacy_array_ptr<char>, legacy_ptr<fc::sha1>) const;
+         void sha512(legacy_array_ptr<char>, legacy_ptr<fc::sha512>) const;
+         void ripemd160(legacy_array_ptr<char>, legacy_ptr<fc::ripemd160>) const;
+
+         REGISTER_LEGACY_CF_HOST_FUNCTION(assert_recover_key);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(recover_key);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(assert_sha256);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(assert_sha1);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(assert_sha512);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(assert_ripemd160);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(sha256);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(sha1);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(sha512);
+         REGISTER_LEGACY_CF_HOST_FUNCTION(ripemd160);
+
+         // permission api
+         bool check_transaction_authorization(legacy_array_ptr<char>, legacy_array_ptr<char>, legacy_array_ptr<char>);
+         bool check_permission_authorization(account_name, permission_name, legacy_array_ptr<char>, legacy_array_ptr<char>, uint64_t);
+         int64_t get_permission_last_used(account_name, permission_name) const;
+         int64_t get_account_creation_time(account_name) const;
+
+         REGISTER_LEGACY_HOST_FUNCTION(check_transaction_authorization);
+         REGISTER_LEGACY_HOST_FUNCTION(check_permission_authorization);
+         REGISTER_HOST_FUNCTION(get_permission_last_used);
+         REGISTER_HOST_FUNCTION(get_account_creation_time);
+
+         // authorization api
+         void require_auth(account_name);
+         void require_auth2(account_name, permission_name);
+         bool has_auth(account_name) const;
+         void require_recipient(account_name);
+         bool is_account(account_name) const;
+
+         REGISTER_HOST_FUNCTION(require_auth);
+         REGISTER_HOST_FUNCTION(require_auth2);
+         REGISTER_HOST_FUNCTION(has_auth);
+         REGISTER_HOST_FUNCTION(require_recipient);
+         REGISTER_HOST_FUNCTION(is_account);
+
+         // system api
+         uint64_t current_time() const;
+         uint64_t publication_time() const;
+         bool is_feature_activated(const digest_type&) const;
+         name get_sender() const;
+
+         REGISTER_HOST_FUNCTION(current_time);
+         REGISTER_HOST_FUNCTION(publication_time);
+         REGISTER_HOST_FUNCTION(is_feature_activated);
```

Comment:
This should be legacy. We don't want to support `const digest_type&` as an argument for host functions going forward.

Also, I'm concerned about the possible changed behavior of references. I think we may need a new `legacy_ref` wrapper type and make this host function's argument type into `legacy_ref<const digest_type>`.

Reply:
No need for `legacy_ref`. This argument would type would instead be `legacy_ptr<const digest_type>`.